src/core/server/saved_objects/deprecations/unknown_object_types.ts

```diff
@@ -0,0 +1,168 @@
+import type { KibanaConfigType } from '../../kibana_config';
+import type { SavedObjectConfig } from '../saved_objects_config';
+import { getIndexForType } from '../service/lib';
+
+interface UnknownTypesDeprecationOptions {
+  typeRegistry: ISavedObjectTypeRegistry;
+  esClient: IScopedClusterClient;
+  kibanaConfig: KibanaConfigType;
+  savedObjectsConfig: SavedObjectConfig;
+  kibanaVersion: string;
+}
+
+const getKnownTypes = (typeRegistry: ISavedObjectTypeRegistry) =>
+  typeRegistry.getAllTypes().map((type) => type.name);
+
+const getTargetIndices = ({
+  types,
+  typeRegistry,
+  kibanaVersion,
+  kibanaConfig,
+  savedObjectsConfig,
+}: {
+  types: string[];
+  typeRegistry: ISavedObjectTypeRegistry;
+  savedObjectsConfig: SavedObjectConfig;
+  kibanaConfig: KibanaConfigType;
+  kibanaVersion: string;
+}) => {
+  return [
+    ...new Set(
+      types.map((type) =>
+        getIndexForType({
+          type,
+          typeRegistry,
+          migV2Enabled: savedObjectsConfig.migration.enableV2,
+          kibanaVersion,
+          defaultIndex: kibanaConfig.index,
+        })
+      )
+    ),
+  ];
+};
+
+const getUnknownTypesQuery = (knownTypes: string[]): estypes.QueryDslQueryContainer => {
+  return {
+    bool: {
+      must_not: knownTypes.map((type) => ({
+        term: { type },
+      })),
+    },
+  };
+};
+
+const getUnknownSavedObjects = async ({
+  typeRegistry,
+  esClient,
+  kibanaConfig,
+  savedObjectsConfig,
+  kibanaVersion,
+}: UnknownTypesDeprecationOptions) => {
+  const knownTypes = getKnownTypes(typeRegistry);
+  const targetIndices = getTargetIndices({
+    types: knownTypes,
+    typeRegistry,
+    kibanaConfig,
+    kibanaVersion,
+    savedObjectsConfig,
+  });
+  const query = getUnknownTypesQuery(knownTypes);
+
+  const { body } = await esClient.asInternalUser.search<SavedObjectsRawDocSource>({
+    index: targetIndices,
+    body: {
+      size: 10000,
+      query,
+    },
+  });
+  const { hits: unknownDocs } = body.hits;
+
+  return unknownDocs.map((doc) => ({ id: doc._id, type: doc._source?.type ?? 'unknown' }));
+};
+
+export const getUnknownTypesDeprecations = async (
+  options: UnknownTypesDeprecationOptions
+): Promise<DeprecationsDetails[]> => {
+  const deprecations: DeprecationsDetails[] = [];
+  const unknownDocs = await getUnknownSavedObjects(options);
+  if (unknownDocs.length) {
+    deprecations.push({
+      title: i18n.translate('core.savedObjects.deprecations.unknownTypes.title', {
+        defaultMessage: 'Saved objects with unknown types are present in Kibana system indices',
+      }),
+      message: i18n.translate('core.savedObjects.deprecations.unknownTypes.message', {
+        defaultMessage:
+          'Upgrades will fail for 8.0+ because documents were found for unknown saved object types.' +
+          `To ensure that upgrades will succeed in the future, either re-enable plugins or delete these documents from the Kibana indices`,
+      }),
```

Comment:
Should we surface the number of unknown docs, of their types, in the deprecation's `message`?

Reply:
In the deprecation guide we recommend not writing `upgrade will fail in version x.x`

Reply:
If we already have the number of unknown docs i think it makes sense to surface them here.